p4-16/spec/P4-16-draft-spec.mdk

```diff
@@ -3363,6 +3363,39 @@ manipulate the elements at the front and back of the stack:
   be a positive integer that is a compile-time known value. The return
   type is ```void```.
 
+The pseudocode below defines the behavior of ```push_front``` and ```pop_front```.
+
+~ Begin P4Pseudo
+void push_front(int count) {
+    for (int i = this.size-1; i >= 0; i -= 1) {
+        if (i >= count) {
+            this[i] = this[i-count];
+        } else {
+            this[i].setInvalid();
+        }
+    }
+    this.nextIndex = this.nextIndex + count;
+    if (this.nextIndex > this.size) this.nextIndex = this.size;
+    // Note: this.last, this.next, and this.lastIndex adjust with this.nextIndex
```

Comment:
There is not a separate data value, but the spec mentions that lastIndex is nextIndex-1.  My comment is there simply to ensure that if some implementer decides to make the .next .last and .lastIndex as separate data items, they should be adjusted as and when nextIndex is updated.